File: zul/src/org/zkoss/zul/Row.java
/* Row.java

	Purpose:
		
	Description:
		
	History:
		Tue Oct 25 16:02:43     2005, Created by tomyeh

Copyright (C) 2005 Potix Corporation. All Rights Reserved.

{{IS_RIGHT
	This program is distributed under LGPL Version 2.1 in the hope that
	it will be useful, but WITHOUT ANY WARRANTY.
}}IS_RIGHT
*/
package org.zkoss.zul;

import java.io.Serializable;
import java.util.Iterator;

import org.zkoss.lang.Objects;
import org.zkoss.zk.ui.Component;
import org.zkoss.zk.ui.UiException;
import org.zkoss.zk.ui.WrongValueException;
import org.zkoss.zul.impl.LoadStatus;
import org.zkoss.zul.impl.XulElement;

/**
 * A single row in a {@link Rows} element.
 * Each child of the {@link Row} element is placed in each successive cell
 * of the grid. The row with the most child elements determines the number
 * of columns in each row.
 *
 * <p>Default {@link #getZclass}: z-row. (since 3.5.0)
 *
 * @author tomyeh
 */
public class Row extends XulElement {
	private static final long serialVersionUID = 20091111L;

	private transient Object _value;
	private String _align, _valign;
	private String _spans;
	private boolean _nowrap;
	/** whether the content of this row is loaded; used if
	 * the grid owning this row is using a list model.
	 */
	private boolean _loaded;
	
	private transient Detail _detail;
	private transient int _index = -1;

	/**
	 * Returns the child detail component.
	 * @since 3.5.0
	 */
	public Detail getDetailChild() {
		return _detail;
	}
	/** Returns the grid that contains this row. */
	public Grid getGrid() {
		final Component parent = getParent();
		return parent != null ? (Grid)parent.getParent(): null;
	}

	/** Returns the horizontal alignment of the whole row.
	 * <p>Default: null (system default: left unless CSS specified).
	 */
	public String getAlign() {
		return _align;
	}
	/** Sets the horizontal alignment of the whole row.
	 * Allowed values: right, left, center, justify, char
	 */
	public void setAlign(String align) {
		if (!Objects.equals(_align, align)) {
			_align = align;
			smartUpdate("align", _align);
		}
	}
	/** Returns the nowrap.
	 * <p>Default: null (system default: wrap).
	 */
	public boolean isNowrap() {
		return _nowrap;
	}
	/** Sets the nowrap.
	 */
	public void setNowrap(boolean nowrap) {
		if (_nowrap != nowrap) {
			_nowrap = nowrap;
			smartUpdate("nowrap", _nowrap);
		}
	}
	/** Returns the vertical alignment of the whole row.
	 * <p>Default: null (system default: top).
	 */
	public String getValign() {
		return _valign;
	}
	/** Sets the vertical alignment of the whole row.
	 * Allowed values: top, middle, bottom, baseline
	 */
	public void setValign(String valign) {
		if (!Objects.equals(_valign, valign)) {
			_valign = valign;
			smartUpdate("valign", _valign);
		}
	}
	public boolean setVisible(boolean visible) {
		if (isVisible() != visible) {
			final Rows rows = (Rows) getParent();
			if (rows != null) {
				final Group g = rows.getGroup(getIndex());
				if (g == null || g.isOpen())
					rows.addVisibleItemCount(visible ? 1 : -1);
			}
		}
		return super.setVisible(visible);
	}

	/** Returns the value.
	 * <p>Default: null.
	 * <p>Note: the value is application dependent, you can place
	 * whatever value you want.
	 */
	@SuppressWarnings("unchecked")
	public <T> T getValue() {
		return (T)_value;
	}
	/** Sets the value.
	 * @param value the value.
	 * <p>Note: the value is application dependent, you can place
	 * whatever value you want.
	 */
	public <T> void setValue(T value) {
		_value = value;
	}

	/** Returns the spans, which is a list of numbers separated by comma.
	 *
	 * <p>Default: empty.
	 * @deprecated As of release 5.0.0, use {@link Cell} instead.
	 */
	public String getSpans() {
		return _spans;
	}
	/** Sets the spans, which is a list of numbers separated by comma.
	 *
	 * <p>For example, "1,2,3" means the second column will span two columns
	 * and the following column span three columns, while others occupies
	 * one column.
	 * @deprecated As of release 5.0.0, use {@link Cell} instead.
	 */
	public void setSpans(String spans) throws WrongValueException {
		if (!Objects.equals(spans, _spans)) {
			_spans = spans;
			smartUpdate("spans", spans);
		}
	}


	/** Internal Use only. Sets whether the content of this row is loaded; used if
	 * the grid owning this row is using a list model.
	 */
	/*package*/ void setLoaded(boolean loaded) {
		if (loaded != _loaded) {
			_loaded = loaded;

			final Grid grid = getGrid();
			if (grid != null && grid.getModel() != null)
				smartUpdate("_loaded", _loaded);
		}
	}
	/** Internal Use Only. Returns whether the content of this row is loaded; used if
	 * the grid owning this row is using a list model.
	 */
	/*package*/ boolean isLoaded() {
		return _loaded;
	}
	/** Returns the index of the specified row.
	 * <p>Notice that the performance is not good if there are a lot of rows.
	 * Therefore, this method shall not be used with a huge grid.
	 * @since 6.0.0
	 */
	public int getIndex() {
		return _index;
	}
	
	/*package*/ void setIndexDirectly(int index) {
		_index = index;	
	}
	
	public String getZclass() {
		return _zclass == null ? "z-row" : _zclass;
	}
	
	/**
	 * Returns the group that this row belongs to, or null.
	 * @since 3.5.0
	 */
	public Group getGroup() {
		final Rows rows = (Rows) getParent();
		return (rows != null) ? rows.getGroup(getIndex()) : null;
	}
	/**
	 * @deprecated as of release 6.0.0. To control the size of Grid related 
	 * components, please refer to {@link Grid} and {@link Column} instead.
	 */
	public void setWidth(String width) {
	}
	/**
	 * @deprecated as of release 6.0.0. To control the size of Grid related 
	 * components, please refer to {@link Grid} and {@link Column} instead.
	 */
	public void setHflex(String flex) {
	}

	//-- super --//
	/** Returns the style class.
	 * By default, it is the same as grid's stye class, unless
	 * {@link #setSclass} is called with non-empty value.
	 */
	public String getSclass() {
		final String sclass = super.getSclass();
		if (sclass != null) return sclass;

		final Grid grid = getGrid();
		return grid != null ? grid.getSclass(): sclass;
	}
	
	// super
	protected void renderProperties(org.zkoss.zk.ui.sys.ContentRenderer renderer)
	throws java.io.IOException {
		super.renderProperties(renderer);
		
		render(renderer, "align", _align);
		render(renderer, "valign", _valign);
		render(renderer, "nowrap", _nowrap);
		render(renderer, "spans", _spans);
		render(renderer, "_loaded", _loaded);
		if (_index >= 0)
			renderer.render("_index", _index);
	}
	
	//-- Component --//
	public void beforeParentChanged(Component parent) {
		if (parent != null && !(parent instanceof Rows))
			throw new UiException("Unsupported parent for row: "+parent);
		super.beforeParentChanged(parent);
	}
	
	public void beforeChildAdded(Component newChild, Component refChild) {
		if (newChild instanceof Detail) {
			if (_detail != null && _detail != newChild)
				throw new UiException("Only one detail is allowed: "+this);
		}
		super.beforeChildAdded(newChild, refChild);
	}
	public boolean insertBefore(Component newChild, Component refChild) {
		if (newChild instanceof Detail) {
			//move to the first child
			refChild = getChildren().isEmpty() ? null : getChildren().get(0);
			if (super.insertBefore(newChild, refChild)) {
				_detail = (Detail) newChild;
				return true;
			}
			return false;			
		} else if (refChild != null && refChild == _detail) {
			if (getChildren().size() <= 1) refChild = null;
			else refChild = getChildren().get(1);
		}
		return super.insertBefore(newChild, refChild);
	}

	public void onChildRemoved(Component child) {
		super.onChildRemoved(child);
		if (_detail == child) _detail = null;
	}

	//Cloneable//
	public Object clone() {
		final Row clone = (Row)super.clone();
		if (_detail != null) clone.afterUnmarshal();
		return clone;
	}
	
	private void afterUnmarshal() {
		for (Iterator it = getChildren().iterator(); it.hasNext();) {
			final Object child = it.next();
			if (child instanceof Detail) {
				_detail = (Detail)child;
				break; //done
			}
		}
	}
	// -- Serializable --//
	private synchronized void writeObject(java.io.ObjectOutputStream s)
			throws java.io.IOException {
		s.defaultWriteObject();

		if (_value instanceof Serializable) {
			s.writeBoolean(true);
			s.writeObject(_value);
		} else {
			s.writeBoolean(false);
		}
	}

	private void readObject(java.io.ObjectInputStream s)
	throws java.io.IOException, ClassNotFoundException {
		s.defaultReadObject();
		
		if (s.readBoolean())
			_value = s.readObject();
		
		afterUnmarshal();
	}
	
	//-- ComponentCtrl --//
	public Object getExtraCtrl() {
		return new ExtraCtrl();
	}
	/** A utility class to implement {@link #getExtraCtrl}.
	 * It is used only by component developers.
	 */
	protected class ExtraCtrl extends XulElement.ExtraCtrl
	implements LoadStatus {
		//-- LoadStatus --//
		public boolean isLoaded() {
			return Row.this.isLoaded();
		}

		public void setLoaded(boolean loaded) {
			Row.this.setLoaded(loaded);
		}
		
		public void setIndex(int index) {
			Row.this.setIndexDirectly(index);
		}
	}

}

File: zul/src/main/resources/web/js/zul/grid/Row.ts
/* Row.ts

	Purpose:

	Description:

	History:
		Tue Dec 23 15:26:27     2008, Created by jumperchen

Copyright (C) 2008 Potix Corporation. All Rights Reserved.

This program is distributed under LGPL Version 2.1 in the hope that
it will be useful, but WITHOUT ANY WARRANTY.
*/
var _isPE = (function () {
	var _isPE_ = zk.feature.pe;
	return function () {
		return _isPE_ && zk.isLoaded('zkex.grid');
	};
})();
/**
 * A single row in a {@link Rows} element.
 * Each child of the {@link Row} element is placed in each successive cell
 * of the grid. The row with the most child elements determines the number
 * of columns in each row.
 *
 * @defaultValue {@link getZclass}: z-row.
 *
 */
@zk.WrapClass('zul.grid.Row')
export class Row extends zul.Widget<HTMLTableRowElement> implements zul.mesh.Item {
	override parent?: zul.grid.Rows;
	override nextSibling?: zul.grid.Row;
	override previousSibling?: zul.grid.Row;

	/** @internal */
	_loaded?: boolean;
	/** @internal */
	_index?: number;

	/** @internal */
	_align?: string;
	/** @internal */
	_nowrap?: boolean;
	/** @internal */
	_valign?: string;
	/** @internal */
	_spans?: number[];
	detail?: zkex.grid.Detail;
	/** @internal */
	_musin?: boolean;

	/**
	 * @returns the horizontal alignment of the whole row.
	 * @defaultValue `null` (system default: left unless CSS specified).
	 */
	getAlign(): string | undefined {
		return this._align;
	}

	/**
	 * Sets the horizontal alignment of the whole row.
	 */
	setAlign(align: string, opts?: Record<string, boolean>): this {
		const o = this._align;
		this._align = align;

		if (o !== align || opts?.force) {
			var n = this.$n();
			if (n)
				n.style.textAlign = align;
		}

		return this;
	}

	/**
	 * @returns the nowrap.
	 * @defaultValue `null` (system default: wrap).
	 */
	isNowrap(): boolean {
		return !!this._nowrap;
	}

	/**
	 * Sets the nowrap.
	 */
	setNowrap(nowrap: boolean, opts?: Record<string, boolean>): this {
		const o = this._nowrap;
		this._nowrap = nowrap;

		if (o !== nowrap || opts?.force) {
			var cells = this.$n()?.cells;
			if (cells)
				for (var j = cells.length; j--;)
					cells[j].noWrap = nowrap;
		}

		return this;
	}

	/**
	 * @returns the vertical alignment of the whole row.
	 * @defaultValue `null` (system default: top).
	 */
	getValign(): string | undefined {
		return this._valign;
	}

	/**
	 * Sets the vertical alignment of the whole row.
	 */
	setValign(valign: string, opts?: Record<string, boolean>): this {
		const o = this._valign;
		this._valign = valign;

		if (o !== valign || opts?.force) {
			var n = this.$n();
			if (n)
				n.style.verticalAlign = valign;
		}

		return this;
	}

	/**
	 * @returns the grid that contains this row.
	 */
	getGrid(): zul.grid.Grid | undefined {
		return this.parent ? this.parent.parent : undefined;
	}

	override setVisible(visible: boolean): this {
		if (this.isVisible() != visible) {
			super.setVisible(visible);
			if (this.desktop && this.isStripeable_())
				this.parent!._syncStripe();
		}
		return this;
	}

	/**
	 * @returns the spans, which is a list of numbers separated by comma.
	 * @defaultValue empty.
	 */
	getSpans(): string {
		return zUtl.intsToString(this._spans);
	}

	/**
	 * Sets the spans, which is a list of numbers separated by comma.
	 *
	 * <p>For example, "1,2,3" means the second column will span two columns
	 * and the following column span three columns, while others occupies
	 * one column.
	 */
	setSpans(spans: string): this {
		if (this.getSpans() != spans) {
			this._spans = zUtl.stringToInts(spans, 1);
			this.rerender();
		}
		return this;
	}

	/** @internal */
	_getIndex(): number {
		return this.parent ? this.getChildIndex() : -1;
	}

	/**
	 * @returns the group that this row belongs to, or null.
	 */
	getGroup(): zkex.grid.Group | undefined {
		// TODO: this performance is not good.
		if (_isPE() && this.parent && this.parent.hasGroup())
			// eslint-disable-next-line @typescript-eslint/no-this-alias
			for (var w: zul.grid.Row | undefined = this; w; w = w.previousSibling)
				if (w instanceof zkex.grid.Group) return w;

		return undefined;
	}

	override setStyle(style: string): this {
		if (this._style != style) {
			if (!zk._rowTime) zk._rowTime = jq.now();
			this._style = style;
			this.rerender();
		}
		return this;
	}

	override rerender(skipper?: number | zk.Skipper): this {
		if (this.desktop) {
			super.rerender(skipper);
			if (this.parent)
				this.parent._syncStripe();
		}
		return this;
	}

	override getSclass(): string | undefined {
		var /*safe*/ sclass = super.getSclass();
		if (sclass != null)
			return sclass;

		var grid = this.getGrid();
		return grid ? grid.getSclass() : sclass;
	}

	/** @internal */
	_getChdextr(child: zk.Widget): HTMLElement {
		return child.$n('chdextr') ?? child.$n_();
	}

	override scrollIntoView(): this {
		var bar = this.getGrid()!._scrollbar;
		if (bar) {
			bar.syncSize();
			bar.scrollToElement(this.$n_());
		} else {
			super.scrollIntoView();
		}
		return this;
	}

	/** @internal */
	override insertChildHTML_(child: zk.Widget, before?: zk.Widget, desktop?: zk.Desktop): void {
		var childHTML = this.encloseChildHTML_({
			child: child,
			index: child.getChildIndex(),
			zclass: /*safe*/ this.getZclass()
		})!;
		if (before)
			jq(this._getChdextr(before)).before(/*safe*/ childHTML);
		else
			jq(this).append(/*safe*/ childHTML);
		child.bind(desktop);
	}

	/** @internal */
	override removeChildHTML_(child: zk.Widget, ignoreDom?: boolean): void {
		super.removeChildHTML_(child, ignoreDom);
		jq(child.uuid + '-chdextr', zk).remove();
	}

	/**
	 * Enclose child with HTML tag with TD and DIV,
	 * and return a HTML code or add HTML fragments in out array.
	 * @internal
	 */
	encloseChildHTML_(opts: { out?: string[]; child: zk.Widget; index: number; zclass: string; visible?: boolean }): string | undefined {
		var out = opts.out ?? new zk.Buffer(),
			child = opts.child;

		if (child instanceof zul.wgt.Cell)
			child._headerVisible = opts.visible;
		else {
			out.push('<td id="', /*safe*/ child.uuid, '-chdextr"',
				/*safe*/ this._childAttrs(child, opts.index), ' tabindex="-1" role="gridcell"><div id="', /*safe*/ child.uuid,
				'-cell" class="', zUtl.encodeXML(opts.zclass), '-content">');
		}
		child.redraw(out);
		if (!(child instanceof zul.wgt.Cell))
			out.push('</div></td>');
		if (!opts.out)
			return out.join('');
	}

	/** @internal */
	_childAttrs(child: zk.Widget, index: number): string {
		var realIndex = index, span = 1;
		if (this._spans) {
			for (var j = 0, k = this._spans.length; j < k; ++j) {
				if (j == index) {
					span = this._spans[j];
					break;
				}
				realIndex += this._spans[j] - 1;
			}
		}
		var visible: boolean | undefined, // eslint-disable-line zk/preferStrictBooleanType
			hgh: string | undefined,
			align: string | undefined,
			valign: string | undefined,
			grid = this.getGrid();
		if (grid) {
			var cols = grid.columns;
			if (cols) {
				if (realIndex < cols.nChildren) {
					var col = cols.getChildAt<zul.grid.Column>(realIndex)!;
					visible = col.isVisible();
					hgh = col.getHeight();
					align = col.getAlign();
					valign = col.getValign();
				}
			}
		}
		var /*safe*/ style = this.domStyle_({ visible: true, width: true, height: true }),
			isDetail = zk.isLoaded('zkex.grid') && child instanceof zkex.grid.Detail;
		if (isDetail) {
			var wd = child.getWidth();
			if (wd)
				style += 'width:' + wd + ';';
		}
		if (hgh || align || valign) {
			if (hgh)
				style += 'height:' + hgh + ';';
			if (align)
				style += 'text-align:' + align + ';';
			if (valign)
				style += 'vertical-align:' + valign + ';';
		}
		var clx = isDetail ? child.$s('outer') : this.$s('inner'),
			attrs = '';
		if (span !== 1)
			attrs += ` colspan="${span}"`;
		if (this._nowrap)
			attrs += ' nowrap="nowrap"';
		if (style)
			attrs += ' style="' + style + '"';
		if (visible === false) {
			if (!(zk.isLoaded('zkex.grid') && this instanceof zkex.grid.Group))
				attrs += ' aria-hidden="true"';
			clx += ' ' + this.$s('hidden-column');
		}
		return DOMPurify.sanitize(attrs + ' class="' + clx + '"');
	}

	/**
	 * @returns whether is stripeable or not.
	 * @defaultValue `true`.
	 * @internal
	 */
	isStripeable_(): boolean {
		return true;
	}

	//-- super --//
	/** @internal */
	override domStyle_(no?: zk.DomStyleOptions): string {
		if ((_isPE() && (this instanceof zkex.grid.Group || this instanceof zkex.grid.Groupfoot))
			|| no?.visible)
			return super.domStyle_(no);

		var /*safe*/ style = super.domStyle_(no),
			group = this.getGroup();
		if (this._align)
			style += ' text-align:' + this._align + ';';
		if (this._valign)
			style += ' vertical-align:' + this._valign + ';';
		return group && !group.isOpen() ? style + 'display:none;' : style;
	}

	/** @internal */
	override onChildAdded_(child: zk.Widget): void {
		super.onChildAdded_(child);
		if (zk.isLoaded('zkex.grid') && child instanceof zkex.grid.Detail)
			this.detail = child;
	}

	/** @internal */
	override onChildRemoved_(child: zk.Widget): void {
		super.onChildRemoved_(child);
		if (child == this.detail)
			this.detail = undefined;
	}

	/** @internal */
	override doFocus_(evt: zk.Event): void {
		super.doFocus_(evt);
		//sync frozen
		var grid = this.getGrid(),
			frozen = grid?.frozen,
			tbody = grid?.rows?.$n();
		if (frozen && tbody) {
			const tds = jq(evt.domTarget).parents('td');
			for (var i = 0, j = tds.length; i < j; i++) {
				const td = tds[i];
				if (td.parentNode!.parentNode == tbody) {
					grid!._moveToHidingFocusCell(td.cellIndex);
					break;
				}
			}
		}
	}

	/** @internal */
	override doMouseOver_(evt: zk.Event): void {
		if (this._musin)
			return;
		this._musin = true;
		var n = this.$n();

		// ZK-2250: all children should apply -moz-user-select: none
		if (n && zk.gecko && this._draggable
			&& !jq.nodeName(evt.domTarget, 'input', 'textarea')) {
			jq(n).addClass('z-draggable-over');
		}

		super.doMouseOver_(evt);
	}

	/** @internal */
	override doMouseOut_(evt: zk.Event): void {
		var n = this.$n();
		if (this._musin && jq.isAncestor(n,
			(evt.domEvent as JQuery.MouseOutEvent).relatedTarget as HTMLElement || evt.domEvent!.toElement)) {
			// fixed mouse-over issue for datebox
			this.parent!._musout = this;
			return;
		}
		this._musin = false;

		// ZK-2250: all children should unapply -moz-user-select: none
		if (n && zk.gecko && this._draggable
			&& !jq.nodeName(evt.domTarget, 'input', 'textarea')) {
			jq(n).removeClass('z-draggable-over');
		}

		super.doMouseOut_(evt);
	}

	/** @internal */
	override domClass_(no?: zk.DomClassOptions): string {
		var /*safe*/ cls = super.domClass_(no),
			grid = this.getGrid(),
			sclass: string;
		// NOTE: It has always been the case that the following `this.$n()` possibly
		// returns null or undefined.
		if (grid && jq(this.$n()).hasClass(sclass = grid.getOddRowSclass()))
			return cls + ' ' + sclass;
		return cls;
	}

	/** @internal */
	override deferRedrawHTML_(out: string[]): void {
		out.push('<tr', this.domAttrs_({ domClass: true }), ' class="z-renderdefer"></tr>');
	}

	/** @internal */
	override getFlexContainer_(): HTMLElement | undefined { //use old flex inside tr/td
		return undefined;
	}
}